From 966b529e9fe868f33a2995e2065b6367de64464a Mon Sep 17 00:00:00 2001
From: Ethan Wei <ethanwei@ami.com>
Date: Thu, 4 Jan 2024 19:04:13 -0800
Subject: [PATCH] Add A Delay to Avoid Block_Exception when Create/Delete VLAN
 Interface. Fix VLAN Gateway Empty.

	modified:   src/ethernet_interface.cpp
	modified:   src/network_manager.cpp
---
 src/ethernet_interface.cpp | 31 ++++++++++++++++++++++++++++---
 src/network_manager.cpp    |  5 +++++
 2 files changed, 33 insertions(+), 3 deletions(-)

diff --git a/src/ethernet_interface.cpp b/src/ethernet_interface.cpp
index f9dd330..0f300f1 100644
--- a/src/ethernet_interface.cpp
+++ b/src/ethernet_interface.cpp
@@ -23,6 +23,7 @@
 #include <xyz/openbmc_project/Common/error.hpp>
 
 #include <algorithm>
+#include <chrono>
 #include <filesystem>
 #include <format>
 #include <string>
@@ -1115,6 +1116,9 @@ ObjectPath EthernetInterface::createVLAN(uint16_t id)
                 .c_str()));
     }
 
+    execute(
+        "/bin/systemctl", "systemctl", "stop",
+        fmt::format("phosphor-ipmi-net@{}.service", interfaceName()).c_str());
     auto objRoot = std::string_view(objPath).substr(0, objPath.rfind('/'));
     auto macStr = MacAddressIntf::macAddress();
     std::optional<stdplus::EtherAddr> mac;
@@ -1152,11 +1156,19 @@ ObjectPath EthernetInterface::createVLAN(uint16_t id)
         config::pathForIntfDev(manager.get().getConfDir(), intfName));
 
     manager.get().addReloadPostHook([ifname = interfaceName()]() {
-        execute("/bin/systemctl", "systemctl", "restart",
+        if (ifname.find("usb0") == std::string::npos)
+        {
+            execute(
+                "/bin/systemctl", "systemctl", "reset-failed",
+                fmt::format("phosphor-ipmi-net@{}.service", ifname).c_str());
+            execute(
+                "/bin/systemctl", "systemctl", "start",
                 fmt::format("phosphor-ipmi-net@{}.service", ifname).c_str());
+        }
     });
     writeConfigurationFile();
     manager.get().reloadConfigs();
+    std::this_thread::sleep_for(std::chrono::seconds(3));
 
     return ret;
 }
@@ -1680,6 +1692,9 @@ void EthernetInterface::VlanProperties::delete_()
         if (intf->ifIdx == parentIdx)
         {
             parentIfName = intf->interfaceName();
+            execute("/bin/systemctl", "systemctl", "stop",
+                    fmt::format("phosphor-ipmi-net@{}.service", parentIfName)
+                        .c_str());
             intf->writeConfigurationFile();
         }
     }
@@ -1690,16 +1705,27 @@ void EthernetInterface::VlanProperties::delete_()
         eth.get().manager.get().addReloadPostHook(
             [idx = eth.get().ifIdx]() { system::deleteIntf(idx); });
 
-        eth.get().manager.get().addReloadPostHook([parentIfName]() {
-            execute("/bin/systemctl", "systemctl", "restart",
+        eth.get().manager.get().addReloadPostHook([&, parentIfName]() {
+            if (parentIfName.find("usb0") == std::string::npos)
+            {
+                execute(
+                    "/bin/systemctl", "systemctl", "reset-failed",
+                    fmt::format("phosphor-ipmi-net@{}.service", parentIfName)
+                        .c_str());
+                execute(
+                    "/bin/systemctl", "systemctl", "start",
                     fmt::format("phosphor-ipmi-net@{}.service", parentIfName)
                         .c_str());
+                execute("/bin/networkctl", "networkctl", "reconfigure",
+                        parentIfName.c_str());
+            }
         });
         // Ignore the interface so the reload doesn't re-query it
         eth.get().manager.get().ignoredIntf.emplace(eth.get().ifIdx);
     }
 
     eth.get().manager.get().reloadConfigs();
+    std::this_thread::sleep_for(std::chrono::seconds(3));
 }
 
 void EthernetInterface::reloadConfigs()
diff --git a/src/network_manager.cpp b/src/network_manager.cpp
index 55ff044..bbf813e 100644
--- a/src/network_manager.cpp
+++ b/src/network_manager.cpp
@@ -197,6 +197,11 @@ void Manager::createInterface(const AllIntfInfo& info, bool enabled)
         auto it = interfaces.find(*info.intf.name);
         if (it != interfaces.end())
         {
+            if (interfacesByIdx.find(info.intf.idx) == interfacesByIdx.end())
+            {
+                interfacesByIdx.insert_or_assign(info.intf.idx,
+                                                 it->second.get());
+            }
             it->second->updateInfo(info.intf);
             return;
         }
-- 
2.43.0

